lib/firenest/replicated_state.ex

```diff
@@ -0,0 +1,343 @@
+defmodule Firenest.ReplicatedState do
+  @moduledoc """
+  Distributed key-value store for ephemeral data.
+
+  The key-value pairs are always attached to a lifetime of a
+  process and the state is replicated to all nodes across the
+  topology. When the attached process dies the state will be
+  removed on all nodes and in case the all the state from
+  disconnected nodes will be (temporarily) removed. The state
+  can only be attached to local processes.
+
+  The state is managed through callbacks that are invoked on the
+  node where the process lives and remotely on other nodes when local
+  changes are propagated.
+
+  The state is replicated incrementally through building local "deltas"
+  (changes to the state) and periodically replicating them remotely. Only
+  some amount of recent deltas are retained for "catching up" remote nodes.
+  If remote node is too far behind the current state and a delta-based
+  catch-up can't be performed, the server falls back to transferring the entire
+  local state. If keeping track of incremental changes is not convenient
+  for a particular state type, the value of a delta can be set to be equal
+  to the current state - this will always cause full state transfers.
+  """
```

Comment:
I would add a full working example to the docs or maybe even two (maps and rate limiter).

Reply:
Should it be in here or should we have an `examples` directory? And if we have them as examples - would it actually make sense to release them as components?

Reply:
I would have them inline for now, unless they become too big.